Annotator final polish changes

maciejdrozdowski opened the pull request:
- resolve most TODO fragments from code
- standardize most of common types in Annotator ( preferences -> settings), add ImportObj type

- display metadata in tooltip
- allow to add/edit metadata in annotation Dialog

- clear stash on save
- clear TOOL action on SHAPE unite

- rename CategoryGroup & AnnotationGroup components + adjust props

- add copy annotations action & modal
- add downloadCOCO action
- add missing key actions
- usePolygon - add close action
- add restoreDefaultSettings button in modal
- Standardize textFields in Panels

maciejdrozdowski committed:
Annotator final polish changes

- resolve most TODO fragments from code
- standardize most of common types in Annotator ( preferences -> settings), add ImportObj type

- display metadata in tooltip
- allow to add/edit metadata in annotation Dialog

- clear stash on save
- clear TOOL action on SHAPE unite

- rename CategoryGroup & AnnotationGroup components + adjust props

- add copy annotations action & modal
- add downloadCOCO action
- add missing key actions
- usePolygon - add close action
- add restoreDefaultSettings button in modal
- Standardize textFields in Panels

wojciech-bilicki commented on client/src/Annotator/Info/useDataset.tsx:
```diff
@@ -58,11 +64,29 @@ const useDataset: IUseDataset = imageId => {
     );
 
+    const copyAnnotations = useCallback(
+        async (imageId: number, id: number, categoriesIds: number[]) => {
```
I would say that function taking 3 args should use named args (pass the object into the function)

wojciech-bilicki commented on client/src/Annotator/Info/useInfo.tsx:
```diff
@@ -264,6 +291,46 @@ const useEditor: ISubHook<UseEditorResponse> = (data, setData) => {
+    );
+
+    const editAnnotationMetadata = useCallback(
+        (
```
ditto

wojciech-bilicki commented on client/src/Annotator/Paper/Component/AnnotationGroupComponent.tsx:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Fake component which observe changes in info layer
+ * and react for them in a form of changes in paper.js scope object
```
`observes` `reacts` `to them`

wojciech-bilicki commented on client/src/Annotator/Paper/Component/CategoryGroupComponent.tsx:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Fake component which observe changes in info layer
+ * and react for them in a form of changes in paper.js scope object
```
ditto

wojciech-bilicki commented on client/src/Annotator/Paper/Shape/KeypointShape.ts:
```diff
@@ -13,7 +13,7 @@ class KeypointShape extends paper.Path {
         // Copy segments from Path.Circle into new paper.Path
         // Keypoint has to inherit from paper.Path
         // because is no such class as paper.Path.Circle... only constructor :-(
```
`because there is no`

wojciech-bilicki commented on client/src/Annotator/Paper/Tool/usePolygon.tsx:
```diff
@@ -172,6 +172,8 @@ export const usePolygon: IToolPolygon = (
         cache.current.polygon.removeSegments(length - 1, length);
     }, []);
 
+    const closePath = useCallback(() => _complete(), [_complete])
```
no need to create new function

maciejdrozdowski committed:
PR review fixes + few missing fixes

- Add Config for KeypointShape
- When expand = false - set Selected to false
- set Config.HitTolerance = 3
